### PR TITLE
Brings JPEG2000 endpoints to Facade

### LINKED ISSUE #35: Create facade for jpeg2000 endpoints.
To be done after #33 and #34 

Now that the initial facade and endpoints are in place, we can proceed with creating the facade functions for the JPEG2000 endpoints.

### PR DESCRIPTION
**Pull request by akash5100**
* closes #35

### CODE REVIEW

File: hvpy/api_groups/jpeg2000/tests/test_get_jpx.py [hunk 1]
```diff
@@ -59,20 +55,13 @@ def test_json_response():
...
 
 def test_error_handling():
-    with pytest.raises(ValidationError, match="getJPXInputParameters\nstartTime\n  field required"):
-        getJPXInputParameters(endTime=datetime(2014, 1, 1, 0, 45, 0), sourceId=14)
-
-    with pytest.raises(ValidationError, match="getJPXInputParameters\nendTime\n  field required"):
-        getJPXInputParameters(startTime=datetime(2014, 1, 1, 0, 0, 0), sourceId=14)
-
-    with pytest.raises(ValidationError, match="getJPXInputParameters\nsourceId\n  field required"):
-        getJPXInputParameters(startTime=datetime(2014, 1, 1, 0, 0, 0), endTime=datetime(2014, 1, 1, 0, 45, 0))
```
Comment (nabobalis) [inline hvpy/api_groups/jpeg2000/tests/test_get_jpx.py:75]: Was there a reason to cut this down to 1 check?

File: hvpy/facade.py [hunk 2]
```diff
@@ -27,3 +27,106 @@ def getJP2Image(
...
...JP2GEN_VERSION><HV_JP2GEN_BRANCH_REVISION>No valid revision number found. Bazaar not installed? Using HV_WRITTENBY manually included revision number: 84 [2011/01/10, https://launchpad.net/jp2gen] : % SPAWN: Error managing child process.:  No such file or directory</HV_JP2GEN_BRANCH_REVISION><HV_HVS_DETAILS_FILENAME>hvs_version5.pro</HV_HVS_DETAILS_FILENAME><HV_HVS_DETAILS_FILENAME_VERSION>5.0</HV_HVS_DETAILS_FILENAME_VERSION><HV_COMMENT>JP2 file created locally at Lockheed LMSAL using hv_aia_list2jp2_gs2 at Sat Feb 16 05:06:34 2013.Contact Helioviewer LMSAL Franchise (slater@lmsal.com) for more details/questions/comments regarding this JP2 file.HVS (Helioviewer setup) file used to create this JP2 file: hvs_version5.pro (version 5.0).FITS to JP2 source code provided by ESA/NASA Helioviewer Project [contact the Helioviewer Project at webmaster@helioviewer.org][NASA-GSFC] and is available for download at https://launchpad.net/jp2gen.Please contact the source code providers if you suspect an error in the source code.Full source code for the entire Helioviewer Project can be found at https://launchpad.net/helioviewer.</HV_COMMENT><HV_SUPPORTED>TRUE</HV_SUPPORTED></helioviewer></meta>')'
```
Comment (nabobalis) [inline hvpy/facade.py:46]: We do not want the entire output here.

File: hvpy/facade.py [hunk 3]
```diff
@@ -27,3 +27,106 @@ def getJP2Image(
...
+    """
+    Generate and (optionally) download a custom JPX movie of the specified
+    datasource with one frame per pair of startTimes/endTimes parameters.
+
+    {Insert}
+    Examples
+    --------
+    >>> from hvpy import getJPXClosestToMidPoint
+    >>> from datetime import datetime
+    >>> getJPXClosestToMidPoint(startTimes=[datetime(2014, 1, 1, 0, 0, 0), datetime(2014, 1, 1, 2, 3, 3), datetime(2014, 1, 1, 4, 5, 5)], endTimes=[datetime(2014, 1, 1, 0, 45, 0), datetime(2014, 1, 1, 2, 33, 3), datetime(2014, 1, 1, 4, 54, 5)], sourceId=14, linked=False, verbose=False, jpip=True)
```
Comment (nabobalis) [inline hvpy/facade.py:70]: Cut this down to a list of two.

File: hvpy/facade.py [hunk 4]
```diff
@@ -27,3 +27,106 @@ def getJP2Image(
...
+    images.
+
+    {Insert}
+    Examples
+    --------
+    >>> from hvpy import getStatus
+    >>> getStatus()
+    {'AIA': {'time': '2022-07-11T08:01:35Z', 'level': 1, 'secondsBehind': 1801, 'measurement': 'AIA 94'}, 'COSMO': {'time': '2022-07-05T00:46:07Z', 'level': 4, 'secondsBehind': 546329, 'measurement': 'COSMO KCor'}, 'EIT': {'time': '2013-08-07T13:06:09Z', 'level': 5, 'secondsBehind': 281647527, 'measurement': 'EIT 284'}, 'HMI': {'time': '2022-07-11T06:00:39Z', 'level': 1, 'secondsBehind': 9057, 'measurement': 'HMI Mag'}, 'LASCO': {'time': '2022-07-11T06:30:07Z', 'level': 1, 'secondsBehind': 7289, 'measurement': 'LASCO C3'}, 'MDI': {'time': '2011-01-11T22:39:00Z', 'level': 5, 'secondsBehind': 362742756, 'measurement': 'MDI Int'}, 'SECCHI': {'time': '2022-07-08T23:57:30Z', 'level': 1, 'secondsBehind': 203646, 'measurement': 'EUVI-A 195'}, 'SWAP': {'time': '2022-07-11T06:17:19Z', 'level': 1, 'secondsBehind': 8057, 'measurement': 'SWAP 174'}, 'SXT': {'time': '2001-12-14T21:06:33Z', 'level': 5, 'secondsBehind': 649164303, 'measurement': 'SXT AlMgMn'}, 'XRT': {'time': '2022-06-24T23:56:45Z', 'level': 5, 'secondsBehind': 1413291, 'measurement': 'XRT Any/Any'}}
```
Comment (nabobalis) [inline hvpy/facade.py:129]: We do not want the entire output here either. 

You can use a `...` to truncate the output.

File: hvpy/api_groups/jpeg2000/tests/test_get_jpx.py [hunk 5]
```diff
@@ -59,20 +55,18 @@ def test_json_response():
...
 def test_error_handling():
-    with pytest.raises(ValidationError, match="getJPXInputParameters\nstartTime\n  field required"):
-        getJPXInputParameters(endTime=datetime(2014, 1, 1, 0, 45, 0), sourceId=14)
-
-    with pytest.raises(ValidationError, match="getJPXInputParameters\nendTime\n  field required"):
-        getJPXInputParameters(startTime=datetime(2014, 1, 1, 0, 0, 0), sourceId=14)
+    with pytest.raises(TypeError, match="missing 1 required positional argument: 'startTime'"):
+        getJPX(endTime=datetime(2014, 1, 1, 0, 45, 0), sourceId=14)
+    with pytest.raises(TypeError, match="missing 1 required positional argument: 'endTime'"):
+        getJPX(startTime=datetime(2014, 1, 1, 0, 0, 0), sourceId=14)
```
Comment (nabobalis) [inline hvpy/api_groups/jpeg2000/tests/test_get_jpx.py:67]:
```suggestion
```

File: hvpy/facade.py [hunk 6]
```diff
@@ -27,3 +27,106 @@ def getJP2Image(
...
+    Generate and (optionally) download a custom JPX movie of the specified
+    datasource with one frame per pair of startTimes/endTimes parameters.
+
+    {Insert}
+    Examples
+    --------
+    >>> from hvpy import getJPXClosestToMidPoint
+    >>> from datetime import datetime
+    >>> getJPXClosestToMidPoint(startTimes=[datetime(2014, 1, 1, 0, 0, 0), datetime(2014, 1, 1, 2, 3, 3)], endTimes=[datetime(2014, 1, 1, 0, 45, 0), datetime(2014, 1, 1, 2, 33, 3)], sourceId=14, linked=False, verbose=False, jpip=True)
+    'jpip://helioviewer.org:8090/movies/SDO_AIA_335_F2013-12-31T18.30.00Z_T2013-12-31T21.03.03ZCMP.jpxmid'
```
Comment (nabobalis) [inline hvpy/facade.py:71]:
```suggestion
    'jpip://helioviewer.org:8090/movies/SDO_AIA_335_F2013-12-31T18.30.00Z_T2013-12-31T21...jpxmid'
```
I am unsure if this URL is fixed or will change per call.

File: hvpy/facade.py [hunk 7]
```diff
@@ -27,3 +27,106 @@ def getJP2Image(
...
+    """
+    Generate and (optionally) download a custom JPX movie of the specified
+    datasource with one frame per pair of startTimes/endTimes parameters.
+
+    {Insert}
+    Examples
+    --------
+    >>> from hvpy import getJPXClosestToMidPoint
+    >>> from datetime import datetime
+    >>> getJPXClosestToMidPoint(startTimes=[datetime(2014, 1, 1, 0, 0, 0), datetime(2014, 1, 1, 2, 3, 3)], endTimes=[datetime(2014, 1, 1, 0, 45, 0), datetime(2014, 1, 1, 2, 33, 3)], sourceId=14, linked=False, verbose=False, jpip=True)
```
Comment (nabobalis) [inline hvpy/facade.py:70]: If a keyword is the default value, I think it should be removed from the example.

File: hvpy/facade.py [hunk 8]
```diff
@@ -27,3 +27,106 @@ def getJP2Image(
...
+    """
+    Generate and (optionally) download a custom JPX movie of the specified
+    datasource from the helioviewer.org API.
+
+    {Insert}
+    Examples
+    --------
+    >>> from hvpy import getJPX
+    >>> from datetime import datetime
+    >>> getJPX(startTime=datetime(2014, 1, 1, 0, 0, 0),endTime=datetime(2014, 1, 1, 0, 45, 0),sourceId=14,linked=True,verbose=False,jpip=True,cadence=None)
```
Comment (nabobalis) [inline hvpy/facade.py:103]: If a keyword is the default value, I think it should be removed from the example.

File: hvpy/api_groups/jpeg2000/tests/test_get_jpx_closest_to_mid_point.py [hunk 9]
```diff
@@ -1,57 +1,46 @@
...
 
 def test_error_handling():
-    with pytest.raises(ValidationError, match="getJPXClosestToMidPointInputParameters\nstartTimes\n  field required"):
-        getJPXClosestToMidPointInputParameters(endTimes=endTimes, sourceId=14)
-
-    with pytest.raises(ValidationError, match="getJPXClosestToMidPointInputParameters\nendTimes\n  field required"):
-        getJPXClosestToMidPointInputParameters(startTimes=startTimes, sourceId=14)
-
-    with pytest.raises(ValidationError, match="getJPXClosestToMidPointInputParameters\nsourceId\n  field required"):
-        getJPXClosestToMidPointInputParameters(startTimes=startTimes, endTimes=endTimes)
```
Comment (nabobalis) [inline hvpy/api_groups/jpeg2000/tests/test_get_jpx_closest_to_mid_point.py:54]: Same here with the reduction.

File: hvpy/facade.py (hunk 6 above)
Comment (nabobalis) [inline hvpy/facade.py:71]:
```suggestion
    'jpip://helioviewer.org:8090/movies/SDO_AIA_335_F2014-01-01T00.00.00Z_T2014-01-01T02.33.03ZCMP.jpxmid'
```
The CI says this is the value

File: hvpy/facade.py (hunk 7 above)
Comment (nabobalis) [inline hvpy/facade.py:70]: If a keyword is the default value, I think it should be removed from the example.

File: hvpy/facade.py (hunk 6 above)
Comment (akash5100) [reply hvpy/facade.py:71]: We can do this, but I copy-pasted the value from the python terminal, I am still confused.

File: hvpy/facade.py (hunk 4 above)
Comment (akash5100) [reply hvpy/facade.py:129]: Sure.

Comment (nabobalis) [reply hvpy/facade.py:129]: This hasn't been truncated yet?

File: hvpy/facade.py [hunk 10]
```diff
@@ -27,3 +27,106 @@ def getJP2Image(
...
+    images.
+
+    {Insert}
+    Examples
+    --------
+    >>> from hvpy import getStatus
+    >>> getStatus()
+    {'AIA': {'time': '2022-07-11T08:01:35Z', 'level': 1, 'secondsBehind': 1801, 'measurement': 'AIA 94'}, 'COSMO': {'time': '2022-07-05T00:46:07Z', 'level': 4, 'secondsBehind': 546329, 'measurement': 'COSMO KCor'}, 'EIT': {'time': '2013-08-07T13:06:09Z', 'level': 5, 'secondsBehind': 281647527, 'measurement': 'EIT 284'}, 'HMI': {'time': '2022-07-11T06:00:39Z', 'level': 1, 'secondsBehind': 9057, 'measurement': 'HMI Mag'}, 'LASCO': {'time': '2022-07-11T06:30:07Z', 'level': 1, 'secondsBehind': 7289, 'measurement': 'LASCO C3'}, 'MDI': {'time': '2011-01-11T22:39:00Z', 'level': 5, 'secondsBehind': 362742756, 'measurement': 'MDI Int'}, 'SECCHI': {'time': '2022-07-08T23:57:30Z', 'level': 1, 'secondsBehind': 203646, 'measurement': 'EUVI-A 195'}, 'SWAP': {'time': '2022-07-11T06:17:19Z', 'level': 1, 'secondsBehind': 8057, 'measurement': 'SWAP 174'}, 'SXT': {'time': '2001-12-14T21:06:33Z', 'level': 5, 'secondsBehind': 649164303, 'measurement': 'SXT AlMgMn'}, 'XRT': {'time': '2022-06-24T23:56:45Z', 'level': 5, 'secondsBehind': 1413291, 'measurement': 'XRT Any/Any'}}
```
Comment (nabobalis) [inline hvpy/facade.py:129]:
```suggestion
    {'AIA': {'time': '2022-07-11T12:14:21Z', 'level': 1, 'secondsBehind': 1568, ... 'XRT': {'time': '2022-06-24T23:56:45Z', 'level': 5, 'secondsBehind': 1428224, 'measurement': 'XRT Any/Any'}}
```

**Comment by nabobalis** [comment]
For future reference, you will want to add Fixes to the issue number otherwise the issue won't close on merge and you will want to figure out how to update your branch without having to do a merge commit.

**Comment by nabobalis** [comment]
Looks like the docstrings will need some future changes to get the CI to be happy with it. Have to keep running them locally and changing them to get them to pass.

**Comment by akash5100** [comment]
> For future reference, you will want to add Fixes to the issue number otherwise the issue won't close on merge and you will want to figure out how to update your branch without having to do a merge commit.

The issue number is already linked to this PR.

**Comment by nabobalis** [comment]
> > For future reference, you will want to add Fixes to the issue number otherwise the issue won't close on merge and you will want to figure out how to update your branch without having to do a merge commit.
> 
> The issue number is already linked to this PR.

Oh they have changed that, that never used to work like that.

File: hvpy/facade.py [hunk 11]
```diff
@@ -27,3 +27,106 @@ def getJP2Image(
...
+    """
+    Returns information about how far behind the latest available JPEG2000
+    images.
+
+    {Insert}
+    Examples
+    --------
+    >>> from hvpy import getStatus
+    >>> getStatus()
+    {'AIA': {'time': '2022-07-11T15:35:26Z', 'level': 1, 'secondsBehind': 2525, 'measurement': 'AIA 1600'}, ... 'XRT': {'time': '2022-06-24T23:56:45Z', 'level': 5, 'secondsBehind': 1441246, 'measurement': 'XRT Any/Any'}}
```
Comment (nabobalis) [inline hvpy/facade.py:129]: I think these numbers are going to change everytime you run the status, you might want to just add `...` for the ones that do.

File: hvpy/facade.py [hunk 12]
```diff
@@ -27,3 +27,106 @@ def getJP2Image(
...
+def getStatus() -> Union[bytes, str, Dict[str, Any]]:
+    """
+    Returns information about how far behind the latest available JPEG2000
+    images.
+
+    {Insert}
+    Examples
+    --------
+    >>> from hvpy import getStatus
+    >>> getStatus() #doctest: +ELLIPSIS
```
Comment (akash5100) [inline hvpy/facade.py:128]: I found the solution here: http://pymotw.com/2/doctest/#handling-unpredictable-output

File: hvpy/facade.py [hunk 13]
```diff
@@ -27,3 +27,106 @@ def getJP2Image(
...
+) -> Union[bytes, str, Dict[str, Any]]:
+    """
+    Get the XML header embedded in a JPEG2000 image. Includes the FITS header
+    as well as a section of Helioviewer-specific metadata.
+
+    {Insert}
+    Examples
+    --------
+    >>> from hvpy import getJP2Header
+    >>> getJP2Header(id=9838343,callback="xml_header") #doctest: +ELLIPSIS
```
Comment (nabobalis) [inline hvpy/facade.py:45]:
```suggestion
    >>> getJP2Header(id=9838343,callback="xml_header")
```
This should be already setup.

File: hvpy/facade.py [hunk 14]
```diff
@@ -27,3 +27,106 @@ def getJP2Image(
...
+    """
+    Generate and (optionally) download a custom JPX movie of the specified
+    datasource with one frame per pair of startTimes/endTimes parameters.
+
+    {Insert}
+    Examples
+    --------
+    >>> from hvpy import getJPXClosestToMidPoint
+    >>> from datetime import datetime
+    >>> getJPXClosestToMidPoint(startTimes=[datetime(2014, 1, 1, 0, 0, 0), datetime(2014, 1, 1, 2, 3, 3)], endTimes=[datetime(2014, 1, 1, 0, 45, 0), datetime(2014, 1, 1, 2, 33, 3)], sourceId=14, linked=False, jpip=True) #doctest: +ELLIPSIS
```
Comment (nabobalis) [inline hvpy/facade.py:70]:
```suggestion
    >>> getJPXClosestToMidPoint(startTimes=[datetime(2014, 1, 1, 0, 0, 0), datetime(2014, 1, 1, 2, 3, 3)], endTimes=[datetime(2014, 1, 1, 0, 45, 0), datetime(2014, 1, 1, 2, 33, 3)], sourceId=14, linked=False, jpip=True)
```
This should be already setup.

You will want want to line break this example, its a bit too long.

File: hvpy/facade.py [hunk 15]
```diff
@@ -27,3 +27,106 @@ def getJP2Image(
...
+def getStatus() -> Union[bytes, str, Dict[str, Any]]:
+    """
+    Returns information about how far behind the latest available JPEG2000
+    images.
+
+    {Insert}
+    Examples
+    --------
+    >>> from hvpy import getStatus
+    >>> getStatus() #doctest: +ELLIPSIS
```
Comment (nabobalis) [inline hvpy/facade.py:128]:
```suggestion
    >>> getStatus()
```
This should be already setup.

**Review by nabobalis — APPROVED** [review]

**Review by dgarciabriseno — APPROVED** [review]

**Pull request merged**